### ISSUE TITLE
"Shape [5400000,5400000] is too large" Error using a large number of dimentions; possible embedding solution for non-NLP task? 

### ISSUE THREAD

**Issue by bbogart** [fchollet/keras#4986]
I have a very large number of dimensions in my training data (5.4e06), which is very sparse. I wanted to try using the data raw as I could not find any embedding examples for continuous non-NLP tasks. For background on the problem see [this question](http://stackoverflow.com/questions/40428432/vector-representation-in-multidimentional-time-series-prediction-in-tensorflow).

I've modified the code for my toy case (#4870), to use a subset of my real data as follows. The 10 sample npy training data is available [here](http://www.ekran.org/tmp/clusterProbability.npy) (~50MB).

```

#!/usr/bin/python
# This program leans to model sequences using an RNN (LSTM network)

import numpy as np

from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM

# Input sequence
wholeSequence = np.load("ML_data/clusterProbability.npy")

# Preprocess Data:
data = wholeSequence[:-1] # all but last
target = wholeSequence[1:] # all but first

# Data Params
numDims = data.shape[1]
numSamples = data.shape[0]

# Reshape training data for Keras LSTM model
# The training data needs to be (batchIndex, timeStepIndex, dimentionIndex)
data = data.reshape((1, numSamples, numDims))
target = target.reshape((1, numSamples, numDims))

# Build Model
model = Sequential()  
model.add(LSTM(numDims, input_shape=(numSamples, numDims), unroll=True, return_sequences=True)) 
model.add(Dense(numDims))
model.add(Dense(numDims))
model.compile(loss='mean_absolute_error', optimizer='adam', metrics=['mean_squared_error'])
model.fit(data, target, nb_epoch=10, batch_size=1, verbose=2)

# Save model to disk.
model.save("sequenceModel.h5")
```

Unfortunately I'm unable to fit the model because of the following error: 

```
Using TensorFlow backend.
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:111] successfully opened CUDA library libcurand.so locally
I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:925] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
I tensorflow/core/common_runtime/gpu/gpu_device.cc:951] Found device 0 with properties: 
name: GeForce GTX 780
major: 3 minor: 5 memoryClockRate (GHz) 0.941
pciBusID 0000:01:00.0
Total memory: 2.95GiB
Free memory: 2.85GiB
I tensorflow/core/common_runtime/gpu/gpu_device.cc:972] DMA: 0 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] 0:   Y 
I tensorflow/core/common_runtime/gpu/gpu_device.cc:1041] Creating TensorFlow device (/gpu:0) -> (device: 0, name: GeForce GTX 780, pci bus id: 0000:01:00.0)
W tensorflow/core/framework/op_kernel.cc:958] Invalid argument: Cannot parse tensor from proto: dtype: DT_FLOAT
tensor_shape {
  dim {
    size: 5400000
  }
  dim {
    size: 5400000
  }
}
float_val: 0

E tensorflow/core/common_runtime/executor.cc:334] Executor failed to create kernel. Invalid argument: Cannot parse tensor from proto: dtype: DT_FLOAT
tensor_shape {
  dim {
    size: 5400000
  }
  dim {
    size: 5400000
  }
}
float_val: 0

	 [[Node: Const_92 = Const[dtype=DT_FLOAT, value=<Invalid TensorProto: dtype: DT_FLOAT tensor_shape { dim { size: 5400000 } dim { size: 5400000 } } float_val: 0>, _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
W tensorflow/core/framework/op_kernel.cc:958] Invalid argument: Cannot parse tensor from proto: dtype: DT_FLOAT
tensor_shape {
  dim {
    size: 5400000
  }
  dim {
    size: 5400000
  }
}
float_val: 0

E tensorflow/core/common_runtime/executor.cc:334] Executor failed to create kernel. Invalid argument: Cannot parse tensor from proto: dtype: DT_FLOAT
tensor_shape {
  dim {
    size: 5400000
  }
  dim {
    size: 5400000
  }
}
float_val: 0

	 [[Node: Const_65 = Const[dtype=DT_FLOAT, value=<Invalid TensorProto: dtype: DT_FLOAT tensor_shape { dim { size: 5400000 } dim { size: 5400000 } } float_val: 0>, _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
W tensorflow/core/framework/op_kernel.cc:958] Invalid argument: Shape [5400000,5400000] is too large (more than 1099511627776 entries)
E tensorflow/core/framework/op_segment.cc:53] Create kernel failed: Invalid argument: Shape [5400000,5400000] is too large (more than 1099511627776 entries)
E tensorflow/core/common_runtime/executor.cc:334] Executor failed to create kernel. Invalid argument: Shape [5400000,5400000] is too large (more than 1099511627776 entries)
	 [[Node: lstm_1_W_i = Variable[container="", dtype=DT_FLOAT, shape=[5400000,5400000], shared_name="", _device="/job:localhost/replica:0/task:0/gpu:0"]()]]
Traceback (most recent call last):
  File "./learnClusterProbability.py", line 41, in <module>
    model.fit(data, target, nb_epoch=10, batch_size=1, verbose=2)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/models.py", line 671, in fit
    initial_epoch=initial_epoch)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/engine/training.py", line 1144, in fit
    initial_epoch=initial_epoch)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/engine/training.py", line 844, in _fit_loop
    outs = f(ins_batch)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/backend/tensorflow_backend.py", line 1601, in __call__
    session = get_session()
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/backend/tensorflow_backend.py", line 119, in get_session
    _initialize_variables()
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/backend/tensorflow_backend.py", line 273, in _initialize_variables
    sess.run(tf.initialize_variables(uninitialized_variables))
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 717, in run
    run_metadata_ptr)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 915, in _run
    feed_dict_string, options, run_metadata)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 965, in _do_run
    target_list, options, run_metadata)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/client/session.py", line 985, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors.InvalidArgumentError: Shape [5400000,5400000] is too large (more than 1099511627776 entries)
	 [[Node: lstm_1_W_i = Variable[container="", dtype=DT_FLOAT, shape=[5400000,5400000], shared_name="", _device="/job:localhost/replica:0/task:0/gpu:0"]()]]

Caused by op u'lstm_1_W_i', defined at:
  File "./learnClusterProbability.py", line 36, in <module>
    model.add(LSTM(numDims, input_shape=(numSamples, numDims), unroll=True, return_sequences=True))
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/models.py", line 298, in add
    layer.create_input_layer(batch_input_shape, input_dtype)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/engine/topology.py", line 398, in create_input_layer
    self(x)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/engine/topology.py", line 543, in __call__
    self.build(input_shapes[0])
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/layers/recurrent.py", line 713, in build
    regularizer=self.W_regularizer)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/engine/topology.py", line 415, in add_weight
    weight = initializer(shape, name=name)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/initializations.py", line 60, in glorot_uniform
    return uniform(shape, s, name=name)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/initializations.py", line 33, in uniform
    return K.random_uniform_variable(shape, -scale, scale, name=name)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/backend/tensorflow_backend.py", line 620, in random_uniform_variable
    return variable(value, dtype=dtype, name=name)
  File "/usr/local/lib/python2.7/dist-packages/Keras-1.2.0-py2.7.egg/keras/backend/tensorflow_backend.py", line 248, in variable
    v = tf.Variable(value, dtype=_convert_string_dtype(dtype), name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variables.py", line 215, in __init__
    dtype=dtype)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/variables.py", line 300, in _init_from_args
    name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/state_ops.py", line 146, in variable_op
    container=container, shared_name=shared_name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/gen_state_ops.py", line 490, in _variable
    name=name)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/op_def_library.py", line 749, in apply_op
    op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 2380, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/framework/ops.py", line 1298, in __init__
    self._traceback = _extract_stack()

InvalidArgumentError (see above for traceback): Shape [5400000,5400000] is too large (more than 1099511627776 entries)
	 [[Node: lstm_1_W_i = Variable[container="", dtype=DT_FLOAT, shape=[5400000,5400000], shared_name="", _device="/job:localhost/replica:0/task:0/gpu:0"]()]]


```

Is this a keras, tensorflow or hardware limitation? 

Some ideas for solutions:

1. **Reduce the dimensionality of my data**; If 1099511627776 max entries is a hard limit, I could rerun my K-means with 38800 clusters (rather than 200,000) but that is a significant reduction of my 30e06 raw samples.

2. **Create a dense representation of my sparse data (i.e. embedding)**; I have no idea how to do this as this is not an NLP task. I don't know what the "vocab size" would be for this data. The data is the probability that each cluster will be in a 2D position for each time-step. Position is represented by a histogram (19 bins for X and 8 bins for Y) for each cluster. There is no meaning in the order of clusters, but there is meaning in the order of the values corresponding to each cluster (e.g. for shape (cluster, position): [[0,1,2], [3, 0, 0]] and [[3, 0, 0], [0,1,2]] are equivalent, but [[1,2,0], [3, 0, 0]] and [[0,1,2], [3, 0, 0]] differ.)

3. **Reshape my data?** As the probabilities are spatial, they could be represented as (19,8) shaped tensors. Would it be any help to organize my data into a 200000 * (19,8) tensors?

Any advice appreciated.

**Comment by bstriner** [comment]
You're just training one sequence. If you only have one training example your model is going to generalize terribly.

Randomly sample sequences of maybe 100 words and train a batch of maybe 32 samples. You could use a generator and fit_generator if you want to randomly generate sequences each epoch. Each batch would be something like (32,100,k).

What is your input? Is is something like a bunch of one-hot vectors? If it is, you can make your model input just the labels and calculate the one-hot vectors on the GPU.

Basically, I'm wondering how you got that many feature vectors, and if that is something you could be doing on the GPU.

**Comment by bstriner** [comment]
Think about it this way: if you only have one sequence, your LSTM can just memorize the answers. If you randomly generate sub-sequences, your LSTM has to learn to predict something based on previous somethings.

Even if you only have one dataset, like one corpus, you want to break it into sub-sequences like sentences or paragraphs to do your training.

**Comment by the-moliver** [comment]
Could you provide more details about the data, what it is and what you want out of it? Based on your description of the raw data (time series data of ~30 million data-points with 5 features) it seems like you are making your life unnecessarily difficult. Why not feed the raw data to a network, batching as necessary?

**Comment by bbogart** [comment]
Thanks @bstriner @the-moliver.

Regarding generalization, I'm more interested in a compressed implicit representation of this sequence (model) than I am in the learning being useful for other sequences. The sequence is highly discontinuous though, so it would have to generalize over sub-sequences. I do plan to do batch training eventually, but as I understand that would not effect the dimensionality of my input. Generalization is not important in this case because I'm more interested in generative properties of the model.

Again, this is not an NLP task, I'll describe how I generated the data in the first place to give you a sense of what it is I'm dealing with. _What I want is a model that reproduces the input sequence (to a low degree of validity) by seeding a feedback loop with the first input vector_. In my toy tests, it seems that even if the model is very well converged, it will still diverge significantly (due to accumulating errors) in the feedback case. This is the code that I have generated such feedback sequences:

```
#!/usr/bin/python

import numpy as np

from keras.models import load_model
from matplotlib import pyplot as plt

# Input sequence
wholeSequence = [[0,0,0,0,0,0,0,0,0,2,1],
                 [0,0,0,0,0,0,0,0,2,1,0],
                 [0,0,0,0,0,0,0,2,1,0,0],
                 [0,0,0,0,0,0,2,1,0,0,0],
                 [0,0,0,0,0,2,1,0,0,0,0],
                 [0,0,0,0,2,1,0,0,0,0,0],
                 [0,0,0,2,1,0,0,0,0,0,0],
                 [0,0,2,1,0,0,0,0,0,0,0],
                 [0,2,1,0,0,0,0,0,0,0,0],
                 [2,1,0,0,0,0,0,0,0,0,0]]

# Preprocess Data:
wholeSequence = np.array(wholeSequence, dtype=float) # Convert to NP array.
data = wholeSequence[:-1] # all but last
data = data.reshape((3, 3, 11))

# Load Model
model = load_model("toySequenceModel.h5")  

# Run a predictive sequence
predict = data # Set seed with initial input data.

# Feedback Loop
sequenceLength = 10
newSequence = np.zeros( (sequenceLength,11),dtype=float )
for i in xrange(sequenceLength):
    predict = model.predict(predict, batch_size=3)

    # Convert float prediction to ints
    predict = np.round(predict, 0)
    predict = predict.astype(int) 

    # Populate new sequence
    newSequence[i,:] = predict[0,0] # Set row i

# Print the sum of each row (sums show the number of percepts per frame!)
print "Sum: ", np.sum(newSequence,axis=1)

# Print the first vector of the prediction:
print newSequence
plt.imshow(newSequence, interpolation='nearest', cmap="gray")
plt.show()
```

### Generating Data

Using a fixed video sequence of 2hours in length as input, I generate a 'vocabulary' of percepts as follows:

1. Segment each frame into regions (~300 per frame), resulting in ~30e06 image segments.

2. To reduce complexity, K-Means reduction of those segments down to 200,000 means (using RGB colour, width and height as features)

3. Create abstractions of regions by calculating the mean of all regions associated with each cluster, resulting in 200,000 images.

This vocabulary is then used to reconstruct the original frames by generating a sequence:

1. For each region segmented above, record the occurrence and positions of each cluster associated with each region for each frame. I want to learn this sequence and reconstruct the frames without having access to the raw data, only the sequence of clusters and cluster positions.

2. The **sequence without positions** would be a vector for 200,000 elements, where 0 indicates non-present at that time-step and 1 if present. This would be one-hot, but the vector would not indicate the number of instances of each cluster for each time-step, nor the position of the clusters.

3. The **sequence with positions and the number of instances** of each cluster for each frame is the sequence I am trying to model. I've reduced the positions of each cluster to histogram with 27bins (19 for X position, 8 for Y). As this is a histogram, it holds the number of instances present in each position bin. 27bins * 200,000 clusters = 5.4e06 dimensions. Using this sequence, I could regenerate the original frames using the number of clusters and their positions. 

### Embedding

If I was using the sequence without positions, then the embedding seems straight forward as the vocabulary size would be 200,000, and the input vector would be one-hot. But then I would not have the number of instances of the positions of the clusters, and be unable to regenerate original frames.

With the sequence including positions, I'm lost as to how I could compress this sparse representation (perhaps an autoencoder/decoder?)

I hope this makes the context more clear, let me know if you have any further questions.

**Comment by the-moliver** [comment]
That does make things more clear but is there a good reason you are doing it this way? Because I work with video data and what you're doing seems bizarre to me. Let me explain. You find clusters in your set of patches and treat them as discrete things. Is there a reason you want that kind of representation? A more powerful representation would be in terms of image features that can take continuous values, but of which you only need a small set to represent a given image patch. This is just sparse coding, and is the basic idea of JPEG compression etc. Representing things as you are, you would have multiple clusters for what is essentially the same image feature but with different intensity levels. Also 200,000 seems way over-parameterized for image patches. How big are they? 
Finally a convolutional autoencoder could be used to find a set of image features that can reconstruct the frames efficiently. And note that no where near 200,000 filters in the first layer and is needed to reconstruct movie frames. So unless this is some strange academic exercise, there are much better ways to do what you're trying to do.

**Comment by bstriner** [comment]
Performance degrading as the sequence continues is a direct consequence of not batching. Training on many subsequences should give you the same performance across the entire timeseries.

Two approaches to make stuff fit in memory:
* Run minibatches of sequences instead of the entire sequence. Do 32 samples of 32 frames instead of 5m frames and you might be able to fit everything into memory
* Reduce feature space using an autoencoder or something similar

**Comment by bbogart** [comment]
@the-moliver: I understand its a strange approach, but it is developed in the context of a larger (artistic) work and it is important that I have access to individual components in isolation (i.e. clusters represented as individual images). Again, I'm not concerned with generalization / overfitting; 200,000 was somewhat arbitrary, but aesthetically valuable. For better or worse, I'm stuck with this approach for this project. 

@bstriner: As I understand, the error is about the dimensionality of my data, not the number of samples. I can't even do the simplest two sample test without getting the error above. I still don't know where this limit is coming from; keras? HW? tensorflow?

Can either of you suggest how I would add an autoencoder to my architecture above? I suppose it could be something like:

1. Train an autoencoder to generate a dense representation of my sparse data, but if I can't even create a network with 5.4e06 inputs, how can I create an autoencoder? (I tried doing a quick test, but could not find a simple enough autoencoder example to work from.)

2. Use the output of the middle layer (sparse -> dense representation) of the trained autoencoder to provide vectors to my existing LSTM architecture and train.

3. To regenerate samples, use the LSTM predicted output and feed it to the trained autoencoder middle layer (dense -> sparse representation), to recreate the sparse vectors predicted over time.

Do I have this right? Examples appreciated.

**Comment by bstriner** [comment]
One aspect of your error is about dimensionality times batch size.  Likely need to reduce one or the other or both.

The other aspect of the error is about the shape of your network. 
* If you have a dense layer with 5m inputs and 5m outputs, that layer has 5mx5m weights.
* If you have a two layer network with 5m inputs and 5m outputs but k hidden dimensions, that two layer network has 5m*k*2 weights. 

This is what a bottleneck does. As long as your hidden dimension is <2.5m, the two layer model will have fewer parameters than the single layer.

So, you should make your LSTM have few output dimensions. Add a timedistributed dense layer to remap the small output dimension back to 5m outputs. You could also add a timedistributed dense layer to map the large output dimension down to a reasonable dimension.

If you were trying to train an LSTM with 5m output and 5m input dimensions that is definitely why your model is crashing. Something like this could work for you.

```
x = Input((timesteps, input_dim))
h = TimeDistributed(Dense(512))(x)
h = LSTM(512)(x)
h = TimeDistributed(Dense(output_dim))(x)
```

You should be able to run your network if you minibatch sequences and use a bottleneck so your matrices aren't too large, but I can't guarantee it will do anything useful.

**Comment by bbogart** [comment]
thank you @bstriner; I understand the concept here, but I'm confused as to how your code snippet would fit into my existing code (included in the first post above). I understand that this code generates tensors, but I don't understand how they would be used in a model. Could you provide a complete example with compile and fit calls?

**Comment by bstriner** [comment]
Just remember that alternating wide and narrow layers use exponentially less memory than consecutive wide layers.

I was just giving it in functional notation. For example, see if this fits in memory.

```python
model = Sequential()  
model.add(LSTM(256, input_shape=(numSamples, numDims), return_sequences=True)) 
model.add(TimeDistributed(Dense(numDims)))
```

**Comment by bbogart** [comment]
thank you @bstriner; I understand the concept here, but I'm confused as to how your code snippet would fit into my existing code (included in the first post above). I understand that this code generates tensors, but I don't understand how they would be used in a model. Could you provide a complete example with compile and fit calls?

**Comment by bbogart** [comment]
Sorry for the delay @bstriner. 

Your code above runs out of memory if I create an LSTM with more than 2 units! 

How can I calculate how much memory such a network will use? This is a 3GB card, but if I'm only getting 2 units, even a 12GB card will not help all that much. 

I'm thinking about going back to my initial approach of separating the problem into two. One LSTM would learn from a 200,000 element boolean vector to model the presence of clusters for each frame. The other would involve training 200,000 LSTMs (one for each cluster) that would each learn the sequence of position distributions for each frame. Its awkward, but seems like a better approach. 

Any thoughts?

**Comment by wangzhenhui1992** [comment]
I had this problem while I did style change of photos with pretrained VGG16
please see the sencond comment of the issue below
https://github.com/fchollet/keras/issues/8608
